Add support for related fields in generic `IndexView.list_display`

abdlrhman08 opened the pull request:
<!-- Thanks for contributing to Wagtail! 🎉  Please add a description below, explaining the purpose of this pull request - including the issue number of the issue you're fixing (if applicable). -->

Fixes  #11494 

After:
![image](https://github.com/wagtail/wagtail/assets/69439342/7422e282-6f66-4529-84d9-da622219ed06)

Also works for recursive relations

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

lb- commented:
Thanks for this PR @abdlrhman08  I have started the CI and requested a review from @laymonage

laymonage commented on wagtail/admin/views/generic/models.py:
```diff
@@ -256,13 +256,29 @@ def get_buttons(column, instance, *args, **kwargs):
+
+        label, attr = label_for_field(field, model_class, return_attr=True)
+
+        # add the name of relational model to make the label more clear
+        try:
+            label = model_name + " " + label
+        except NameError:
+            pass
```
I don't think this is necessary. Or if we want to, we definitely don't want to use the field's internal name directly here, which is a code string in snake_case and should not be exposed to the user. A better idea would be to use the model's verbose name (`model._meta.verbose_name`)

abdlrhman08 replied:
Changed to model's verbose name, can be removed if not necessary.

laymonage commented on wagtail/admin/views/generic/models.py:
```diff
@@ -256,13 +256,29 @@ def get_buttons(column, instance, *args, **kwargs):
+        # iterate over the relation list to try to get the last model
+        # where the field exists
+        for model in relation:
+            model_class = getattr(model_class, "_meta").get_field(model).related_model
```
Is there a reason why we use `getattr()` here instead of `model_class._meta` when we're always accessing `_meta` and we're not using a default fallback value?

abdlrhman08 replied:
Tbh, I don't remember why I used `getattr()` instead of directly accessing it using `model_class._meta`. I changed it to `model_class._meta`.

laymonage commented on wagtail/admin/views/generic/models.py:
```diff
@@ -256,13 +256,29 @@ def get_buttons(column, instance, *args, **kwargs):
 
     def _get_custom_column(self, field_name, column_class=Column, **kwargs):
-        label, attr = label_for_field(field_name, self.model, return_attr=True)
+        *relation, field = field_name.split(".")
```
Can we split by `django.db.models.constants.LOOKUP_SEP` instead?

abdlrhman08 replied:
Changed

laymonage commented on wagtail/admin/views/generic/models.py:
```diff
@@ -256,13 +256,29 @@ def get_buttons(column, instance, *args, **kwargs):
         # so it's possible to sort by it
         if attr is None:
-            sort_key = field_name
+            sort_key = field_name.replace(".", "_")
```
Does this work? I imagine you'd need double underscores for this to work. If we used `LOOKUP_SEP` instead of dot-separated fields, we probably wouldn't need this.

abdlrhman08 replied:
Yes you are totally right, if `LOOKUP_SEP` is used there is no need for this.

laymonage reviewed (changes requested):
Thanks for giving this a go! I added some thoughts on this feature in https://github.com/wagtail/wagtail/issues/11494#issuecomment-1948769490. Could you have a look (and perhaps also see how Django's ModelAdmin implements it and how it works over there)?

abdlrhman08 commented:
Thank you for your review @laymonage ! I took a look at how Django's ModelAdmin implemented it and, I have changed the code according to requested changes and the tests accordingly. Can you have another look on the changes ? 

There is still one thing I would like you to take a look at, in this [line here](https://github.com/abdlrhman08/wagtail/blob/3758a71b2b425bb17910279b00f905602756a4d5/wagtail/admin/views/generic/models.py#L285), I had to build the dotted path again, because `multigetattr()` uses it in the column to get the data in the field. Should we change `multigetattr` to also split by double underscores or build the dotted field path again as I did.

I am happy to change if any changes are requested !

laymonage commented on wagtail/admin/views/generic/models.py:
```diff
@@ -256,17 +257,35 @@ def get_buttons(column, instance, *args, **kwargs):
+        foreign_field_name = ""
+        for model in relation:
+            foreign_field = model_class._meta.get_field(model)
+            foreign_field_name += foreign_field.verbose_name + " "
```
This can be very long if we traverse several relations, maybe just take the `verbose_name` of the last relation?

```suggestion
            foreign_field_name = foreign_field.verbose_name + " "
```

abdlrhman08 replied:
Yeah, I think taking the last relations name only is enough. I wrote like this in the beginning, because in Django Admin it displays the complete relation

laymonage commented on wagtail/snippets/tests/test_viewset.py:
```diff
@@ -769,6 +769,22 @@ def test_boolean_column(self):
+
+        response = self.client.get(self.get_url("list"))
+
+        self.assertEqual(response.status_code, 200)
```
Could you also assert that the response contains the text of the full featured snippet, as well as the header correctly showing the label (foreign key's verbose_name + the field's verbose_name)

abdlrhman08 replied:
Done

laymonage commented on wagtail/snippets/tests/test_viewset.py:
```diff
@@ -769,6 +769,22 @@ def test_boolean_column(self):
+class TestRelatedFieldColumns(BaseSnippetViewSetTests):
+    model = SnippetChooserModel
+
+    def test_snippetchoosermodel_related_fields(self):
```
Could you add another test case where the foreign key is null? e.g.

```py
    def test_snippetchoosermodel_related_fields(self):
        advert = Advert(url="https://example.com/free_examples", text="Free Examples")
        advert.save()
        scm = SnippetChooserModel(advert=advert)
        scm.save()

        response = self.client.get(self.get_url("list"))

        self.assertEqual(response.status_code, 200)
        # assert the header label here
        # (and maybe assert that the cell is empty)
```

Right now it seems to crash as the `multigetattr` function tries to access `None["text"]`. You may have to add a `try`/`except` in `Column.get_value()` to return `None` when this occurs.

abdlrhman08 replied:
Done

laymonage reviewed (commented):
Thanks, this is looking great and I think it's almost there!

abdlrhman08 commented:
Thank you for your review and help @laymonage ! I added one more test case. At the moment, I can't think of more test cases. I am happy to add any corner cases or more tests.

laymonage commented on wagtail/test/testapp/wagtail_hooks.py:
```diff
@@ -382,12 +383,20 @@ class VariousOnDeleteModelViewSet(SnippetViewSet):
+class SnippetChooserModelViewSet(SnippetViewSet):
+    model = SnippetChooserModel
+
+    list_display = ["__str__", "full_featured__text"]
```
Sorry I forgot to suggest this before, could we add another case for multi-level relations? e.g.

```suggestion
    list_display = ["__str__", "full_featured__text", "full_featured__latest_revision__created_at"]
```

In the test, you'll need to do `ffs.save_revision()` so the `latest_revision` is populated. The method returns the `Revision` object, so you can hold on to it to get the `created_at` for assertion. You can then use `wagtail.utils.timestamps.render_timestamp` to assert the rendered timestamp is in the response. The label should be "Latest revision created at".

laymonage reviewed (commented):
Thank you, this is looking great! One small comment to improve the label for other languages, and another comment to ensure that this feature works with multi-level relations. I think we'll be good to go after that!

abdlrhman08 commented:
I just went through the code of `label_for_field`, and I found out that if we are using `LOOKUP_SEP` in `list_diplay` there is no need for my addition, since `label_for_field` will just work out of the box. I'm sorry for noticing this late. So basically this is enough to do it:
```python
def _get_custom_column(self, field_name, column_class=Column, **kwargs):
        label, attr = label_for_field(field_name, self.model, return_attr=True)
        sort_key = getattr(attr, "admin_order_field", None)

        # attr is None if the field is an actual database field,
        # so it's possible to sort by it
        if attr is None:
            sort_key = field_name

        if not hasattr(self.model, field_name):
            sort_key = field_name
            field_name = field_name.replace(LOOKUP_SEP, ".")
            
        return column_class(
            field_name,
            label=capfirst(label),
            sort_key=sort_key,
            **kwargs,
        )
```

The only problem here is no translation and we will have very long relation name.
let me know what you think about this

laymonage commented on wagtail/admin/views/generic/models.py:
```diff
@@ -256,16 +257,39 @@ def get_buttons(column, instance, *args, **kwargs):
             sort_key = field_name
 
+        # build the dotted relation if needed, for the use in multigetattr
+        dotted_field = ".".join(relation + [field])
```
If we didn't immediately unpack the splitting result into `*relation, field`, we can reuse it here so we don't have to construct a new list. It's a very small optimisation but I think it's nice to have

laymonage commented on wagtail/snippets/tests/test_viewset.py:
```diff
@@ -769,6 +770,47 @@ def test_boolean_column(self):
+    def test_snippetchoosermodel_related_fields(self):
+        response = self.client.get(self.get_url("list"))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, "Full featured text")
```
If we add a `verbose_name` to the `ForeignKey`, we can make a more specific assertion here to ensure that we're using the verbose name of the field rather than just the field name

laymonage commented on wagtail/snippets/tests/test_viewset.py:
```diff
@@ -769,6 +770,47 @@ def test_boolean_column(self):
+            text="test snippet with custom icon"
+        )
+        self.scm = self.model.objects.create(advert=self.advert, full_featured=self.ffs)
+        self.no_ffs_chooser = self.model.objects.create(advert=self.advert)
```
It's best to move the creation of this object to the test for empty FK specifically, as otherwise the other unrelated tests could also fail if the implementation didn't handle the empty FK case.

laymonage commented on wagtail/snippets/tests/test_viewset.py:
```diff
@@ -769,6 +770,47 @@ def test_boolean_column(self):
+        self.scm = self.model.objects.create(advert=self.advert, full_featured=self.ffs)
+        self.no_ffs_chooser = self.model.objects.create(advert=self.advert)
+
+    def test_snippetchoosermodel_related_fields(self):
```
We're not testing the `SnippetChooserModel`, so it's better to use a more descriptive name e.g.

```suggestion
    def test_single_level_relation(self):
```

laymonage reviewed (commented):
Thanks @abdlrhman08, this is looking great! I only have a few minor comments. I'll make the changes as I merge.

> I just went through the code of label_for_field, and I found out that if we are using LOOKUP_SEP in list_diplay there is no need for my addition, since label_for_field would just work out of the box. I'm sorry for noticing this late. So basically this is enough to do it:

No worries, you're right but that only applies to the current in-development version of Django because https://github.com/django/django/pull/17357 is quite recent. If you install the latest release, you'll get an `AttributeError` when trying to do so, e.g.

```
AttributeError: Unable to lookup 'favorite_book__author__name' on Person
```

We currently support Django 4.2 and 5.0, so we still need to provide our own implementation anyway. This also lets us to do some tweaks like only using the final related model, and also adding the ability for other languages to swap the order of the model name vs the field name.

laymonage commented:
Thanks! I've added your name to the contributors list as `Abdelrahman Hamada`. Let me know if you'd like a different name listed.

abdlrhman08 commented:
Thank you ! looking forward to helping the community more
